orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -825,6 +819,206 @@ void AddNewScalarNodeArgAndInitializer(Graph& graph,
   graph.AddInitializedTensor(proto_data);
 }
 
+Status FindAllConnectedNodes(Graph& graph,
```

Comment:
>FindAllConnectedNodes [](start = 7, length = 21)

Maybe better to build the connection with one traverse of the graph for all nodes. Otherwise this is O(n^2) complexity, which would be greatly slow down with big models of thousands of nodes.